tensorflow/compiler/xla/service/all_reduce_reassociate_test.cc

```diff
@@ -332,5 +332,52 @@ ENTRY main {
                           RunPass(hlo_string, /*expect_change=*/true));
 }
 
+// Checks whether a linear chain of converts-adds of ARs is reassociated in a
+// single pass.
+TEST_F(AllReduceSimplifierTest, ChainWithConvert) {
+  absl::string_view hlo_string = R"(
+HloModule m
+
+add.1 {
+  x.47 = bf16[] parameter(0)
+  convert.13139 = f32[] convert(x.47)
+  y.47 = bf16[] parameter(1)
+  convert.13140 = f32[] convert(y.47)
+  add.2532 = f32[] add(convert.13139, convert.13140)
+  ROOT convert.13141 = bf16[] convert(add.2532)
```

Comment:
This is quite confusing; what is the point of pushing conversion into the adder, as opposed to having it after the all-reduce?

Reply:
Since to_apply is launched as a single kernel, doing it this way could potentially leverage the IO benefit of smaller data types. The casts are also auto-inserted by XLA so it might be a bit involved to optimize cross-region casts.

Reply:
Wouldn't it make more sense not to include these casts inside the computation in the first place? NCCL supports bf16 I believe?

Reply:
I agree with @cheshire, the reduction computation attached to all-reduce and reduce-scatter should stay in canonical form and we should not extend the pattern matching to look through casts. We currently do such pattern matching during codegen as well (look for a fusion with casts inside it) but I intend to fix this as well to keep the all-reduce computation in its canonical form and remove the pattern matching in codegen

Reply:
Keeping the allreduce computation in canonical form is fine and that's not what this pr is trying to fix. This is trying to match
             allreduce
                      \
                     (cast)
                            \
allreduce- (cast) - add
and perform the reordering. AFAIK, native gpu bfloat16 add is not available yet in llvm.
The casts inside allreduce compute prevents generation of allreduce keys, using only input data type in adder would make life much easier.
However, I'm not sure how feasible it is to enforce this across the compiler, XLA is aware of storage types right from the start, this means we have to exclude allreduce/reduceScatter from all data-type normalization passes, we can move bfloat16 normalization down for now, it doesn't guarantee the casts won't be inserted for other data types.